Mathlib/Combinatorics/Enumerative/DyckWord.lean

```diff
@@ -77,43 +78,56 @@ instance : AddMonoid DyckWord where
 
 namespace DyckWord
 
-variable (p q : DyckWord)
+variable {p q : DyckWord}
 
 lemma toList_eq_nil : p.toList = [] ↔ p = 0 := by rw [DyckWord.ext_iff]; rfl
+lemma toList_ne_nil : p.toList ≠ [] ↔ p ≠ 0 := toList_eq_nil.ne
+
+@[simp]
+lemma add_eq_zero_iff : p + q = 0 ↔ p = 0 ∧ q = 0 := by
+  simp_rw [← toList_eq_nil]; apply append_eq_nil
+
+lemma add_ne_zero_iff : p + q ≠ 0 ↔ p ≠ 0 ∨ q ≠ 0 := by
+  rw [not_iff_comm, add_eq_zero_iff]; tauto
+
+variable (h : p ≠ 0)
 
 /-- The first element of a nonempty Dyck word is `U`. -/
-lemma head_eq_U (h : p.toList ≠ []) : p.toList.head h = U := by
+lemma head_eq_U (p : DyckWord) (h) : p.toList.head (toList_ne_nil.mpr h) = U := by
   rcases p with - | s; · tauto
   rw [head_cons]
   by_contra f
   rename_i _ nonneg
   simpa [s.dichotomy.resolve_left f] using nonneg 1
 
 /-- The last element of a nonempty Dyck word is `D`. -/
-lemma getLast_eq_D (h : p.toList ≠ []) : p.toList.getLast h = D := by
+lemma getLast_eq_D (p : DyckWord) (h) : p.toList.getLast (toList_ne_nil.mpr h) = D := by
   by_contra f; have s := p.count_U_eq_count_D
-  rw [← dropLast_append_getLast h, (dichotomy _).resolve_right f] at s
+  rw [← dropLast_append_getLast (toList_ne_nil.mpr h), (dichotomy _).resolve_right f] at s
   simp_rw [dropLast_eq_take, count_append, count_singleton', ite_true, ite_false] at s
   have := p.count_D_le_count_U (p.toList.length - 1); omega
 
-lemma cons_tail_dropLast_concat (h : p.toList ≠ []) :
-    U :: p.toList.dropLast.tail ++ [D] = p := by
-  have : p.toList.dropLast.take 1 = [p.toList.head h] := by
+include h in
+lemma cons_tail_dropLast_concat : U :: p.toList.dropLast.tail ++ [D] = p := by
+  have h' := toList_ne_nil.mpr h
+  have : p.toList.dropLast.take 1 = [p.toList.head h'] := by
     rcases p with - | ⟨s, ⟨- | ⟨t, r⟩⟩⟩
     · tauto
     · rename_i bal _
       cases s <;> simp at bal
     · tauto
-  nth_rw 2 [← p.toList.dropLast_append_getLast h, ← p.toList.dropLast.take_append_drop 1]
-  rw [p.getLast_eq_D, drop_one, this, p.head_eq_U]
+  nth_rw 2 [← p.toList.dropLast_append_getLast h', ← p.toList.dropLast.take_append_drop 1]
+  rw [p.getLast_eq_D h, drop_one, this, p.head_eq_U h]
```

Comment:
```suggestion
  rw [p.getLast_eq_D, drop_one, this, p.head_eq_U]
```

Reply:
That doesn't work; the `p ≠ 0` hypotheses are then left as side goals.

Reply:
That's because you haven't removed the `toList_ne_nil.mpr` in the statement of `getLast_eq_D`